src/main/scala/ch/epfl/alcmp/net/Serde.scala

```diff
@@ -22,40 +22,78 @@ object Serde {
+        s + b1.toString
+
+      override def deserialize(data: String): DivideMessage =
+        val arr = data.split("/")
+        val (id, depth, index) = (arr(0).toInt, arr(1).toInt, arr(2).toInt)
+        val highlights = arr(3).split(",").map(s => s.toInt).toList
+        t match
+        case TypeId.ListType =>
+          val outputs = arr(4).split(";")
+                        .map(list => IList(list.split(",").map(s2 => s2.toInt).toList)).toList
+          DivideMessage(id, depth, index, outputs, highlights)
+        case TypeId.MatrixType =>
+          val outputs = arr(4).split(";")
+            .map(lists => IMatrix(lists.split(",,")
+            .map(_.split(",").map(number => number.toInt).toList).toList)).toList
+          DivideMessage(id, depth, index, outputs, highlights)
+        case TypeId.HeapType =>
+          val outputs = arr(4).split(";")
+            .map(list => IHeap(list.split(",").map(s2 => s2.toInt).toList)).toList
+          DivideMessage(id, depth, index, outputs, highlights)
         case TypeId.BinaryTreeType => ???
 
   given (TypeId => Serde[CombineMessage]) with
     override def apply(t: TypeId): Serde[CombineMessage] = new Serde[CombineMessage]:
-      override def serialize(obj: CombineMessage): String = obj.output match {
-        case IList(list) => ???
-        case IMatrix(rows) => ???
-        case IHeap(list) => ???
-        case IBinaryTree(root) => ???
-      }
-      override def deserialize(data: String): CombineMessage = t match
-        case TypeId.ListType => ???
-        case TypeId.MatrixType => ???
-        case TypeId.HeapType => ???
+      override def serialize(obj: CombineMessage): String =
+        val commonPart = List(obj.id, obj.depth, obj.index, obj.highlights.mkString(",")).mkString("/")
+        obj.output match {
+          case IList(list) => commonPart + "/" + list.mkString(",")
+          case IMatrix(rows) =>
+            val b = StringJoiner(",,")
+            rows.foreach(list => b.add(list.mkString(",")))
+            commonPart + "/" + b.toString
+          case IHeap(list) => commonPart + '/' + list.mkString(",")
+          case IBinaryTree(root) => ???
+        }
+      override def deserialize(data: String): CombineMessage =
+        val arr = data.split("/")
+        val (id, depth, index) = (arr(0).toInt, arr(1).toInt, arr(2).toInt)
+        val highlights = arr(3).split(",").map(s => s.toInt).toList
+        t match
+        case TypeId.ListType =>
+          val output = arr(4).split(",").map(s => s.toInt).toList
+          CombineMessage(id, depth, index, IList(output), highlights)
+        case TypeId.MatrixType =>
+          val output = arr(4).split(",,").map(s => s.split(",").map(s => s.toInt).toList).toList
+          CombineMessage(id, depth, index, IMatrix(output), highlights)
+        case TypeId.HeapType =>
+          val output = arr(4).split(",").map(s => s.toInt).toList
+          CombineMessage(id, depth, index, IHeap(output), highlights)
```

Comment:
Still the same thing, I think we could improve the code by making it less "protocol-specific" using composition